src/common_modules/vsts-cli-package-common/vsts/cli/package/common/artifacttool_updater.py

```diff
@@ -0,0 +1,137 @@
+        logger.debug("Looking up current version of ArtifactTool...")
+        release = client.get_clienttool_release("ArtifactTool", os_name=platform.system(), arch=platform.machine(), version=override_version)
+        return (release.uri, self._compute_id(release)) if release is not None else None
+
+    def _update_artifacttool(self, uri, release_dir, release_id):
+        with humanfriendly.Spinner(label="Downloading UPack tooling (%s)" % release_id, total=100, stream=sys.stderr) as spinner:
+            spinner.step()
+            logger.debug("Downloading ArtifactTool from %s" % uri)
+
+            # Make the request, determine the total size
+            response = requests.get(uri, stream=True)
+            content_length_header = response.headers['Content-Length'].strip()
+            content_length = int(content_length_header)
+
+            # Do the download, updating the progress bar
+            content=io.BytesIO()
+            bytes_so_far = 0
+            for chunk in response.iter_content(chunk_size=1024*512):
+                if chunk:
+                    content.write(chunk)
+                    bytes_so_far += len(chunk)
+                    spinner.step(100 * float(bytes_so_far)/float(content_length))
+
+            # Extract the zip
+            release_temp_dir = self._get_temp_release_dir()
+            logger.debug("Extracting ArtifactTool to %s" % release_temp_dir)
+            f = zipfile.ZipFile(content)
+            f.extractall(path=release_temp_dir)
+
+            # For Linux, ensure the executable bit is set on the binary "ArtifactTool" if it exists.
+            # Python has a bug https://bugs.python.org/issue15795 where file permissions are not preserved.
+            artifacttool_binary = os.path.join(release_temp_dir, "ArtifactTool")
+            if os.path.exists(artifacttool_binary):
+                artifacttool_stat = os.stat(artifacttool_binary)
+                os.chmod(artifacttool_binary, artifacttool_stat.st_mode | stat.S_IXUSR | stat.S_IXGRP | stat.S_IXOTH)
+
+            # Move the directory into the real location
+            logger.debug("Moving downloaded ArtifactTool from %s to %s" % (release_temp_dir, release_dir))
+            os.rename(release_temp_dir, release_dir)
+            logger.info("Downloaded UPack tooling successfully")
+
+    def _mkdir_if_not_exist(self, path):
+        try: 
+            os.makedirs(path)
+        except OSError:
+            # Ignore errors that were likely because the directory already exists
+            if not os.path.isdir(path):
+                raise
+
+    def _compute_id(self, release):
+        return "%s_%s_%s" % (release.name, release.rid, release.version)
+
+    def _compute_artifact_root(self):
+        temp_dir = tempfile.gettempdir()
+        return os.path.join(temp_dir, "ArtifactTool")
+
+    def _get_release_dir(self, release_id):
```

Comment:
You may want to consider a cleanup mechanism for older versions that are no longer used.  Not sure how many different versions a user would use, but looks like this is just set up to grow as new versions are needed.

Reply:
Agree that we need this at some point. I'm happy to build it in now if we think that ArtifactTool will release frequently before we'd have a chance to implement and release this in the future. Mitch wanted the ability to pin to different versions, so I don't think the cleanup mechanism would be as simple as "keep latest". Let me see what Mitch thinks.